File: app/src/main/java/com/simple/tracking/admin/fragment/DeliveryFragment.java
package com.simple.tracking.admin.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.PermissionRequest;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.simple.tracking.LocationChecker;
import com.simple.tracking.PreferenceManager;
import com.simple.tracking.R;
import com.simple.tracking.admin.activity.delivery.AdminCreateDeliveryDetailActivity;
import com.simple.tracking.admin.adapter.DeliveryAdapter;
import com.simple.tracking.model.Delivery;
import com.simple.tracking.network.BaseResponse;
import com.simple.tracking.network.DeliveryAPIConfiguration;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import java.util.List;
import java.util.Objects;

public class DeliveryFragment extends Fragment {

    private DeliveryAdapter deliveryAdapter;
    private RecyclerView recyclerView;
    private TextInputEditText searchInput;
    private int currentPage = 1;
    private boolean isLastPage = false;
    private boolean isLoading = false;
    private final int PAGE_SIZE = 10;
    private final Handler searchHandler = new Handler();
    private Runnable searchRunnable;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_admin_delivery, container, false);

        if (!LocationChecker.isLocationEnabled(requireActivity())) {
            LocationChecker.showLocationDisabledDialog(requireActivity());
        }

        recyclerView = view.findViewById(R.id.recyclerViewDelivery);
        searchInput = view.findViewById(R.id.search_input_delivery);
        MaterialButton btnSearch = view.findViewById(R.id.btn_search_delivery);
        CardView btnAddDelivery = view.findViewById(R.id.btn_add_delivery);

        PreferenceManager preferenceManager = new PreferenceManager(requireContext());
        if (preferenceManager.isAdmin()) btnAddDelivery.setVisibility(View.VISIBLE);

        btnAddDelivery.setOnClickListener(v -> {
            Intent deliveryCreate = new Intent(requireContext(), AdminCreateDeliveryDetailActivity.class);
            deliveryCreate.putExtra("MENU_NAME", "Delivery");
            startActivity(deliveryCreate);
        });

        getDeliveries(null);

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
                if (layoutManager != null && !isLoading && !isLastPage) {
                    int visibleItemCount = layoutManager.getChildCount();
                    int totalItemCount = layoutManager.getItemCount();
                    int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();

                    if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount
                            && firstVisibleItemPosition >= 0
                            && totalItemCount >= PAGE_SIZE) {
                        getDeliveries(null);
                    }
                }
            }
        });

        btnSearch.setOnClickListener(v -> {
            currentPage = 1;
            isLastPage = false;
            isLoading = false;

            if (deliveryAdapter != null) {
                deliveryAdapter.clearDeliveries();
            }

            if (searchInput == null || searchInput.toString().trim().isEmpty()) {
                getDeliveries(null);
            } else {
                getDeliveries(Objects.requireNonNull(searchInput.getText()).toString());
            }
        });

        searchInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) { }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) { }

            @Override
            public void afterTextChanged(Editable s) {
                if (searchRunnable != null) {
                    searchHandler.removeCallbacks(searchRunnable);
                }

                searchRunnable = () -> {
                    if (s.toString().isEmpty()) {
                        currentPage = 1;
                        isLastPage = false;
                        isLoading = false;

                        if (deliveryAdapter != null) {
                            deliveryAdapter.clearDeliveries();
                        }

                        getDeliveries(null);
                    }
                };

                searchHandler.postDelayed(searchRunnable, 1000);
            }
        });

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();

        if (!LocationChecker.isLocationEnabled(requireActivity())) {
            LocationChecker.showLocationDisabledDialog(requireActivity());
        }

        currentPage = 1;
        isLastPage = false;

        getDeliveries(null);
    }

    private void getDeliveries(String query) {
        if (isLoading || isLastPage) return;

        isLoading = true;

        Call<BaseResponse<List<Delivery>>> call;

        PreferenceManager preferenceManager = new PreferenceManager(requireContext());
        if (preferenceManager.isAdmin()) {
            call = DeliveryAPIConfiguration.getInstance().getDeliveries(query, currentPage, PAGE_SIZE);
        } else {
            call = DeliveryAPIConfiguration.getInstance().getDeliveriesShipper(preferenceManager.getUserId(), query, currentPage, PAGE_SIZE);
        }

        call.enqueue(new Callback<BaseResponse<List<Delivery>>>() {
            @Override
            public void onResponse(@NonNull Call<BaseResponse<List<Delivery>>> call, @NonNull Response<BaseResponse<List<Delivery>>> response) {
                isLoading = false;
                if (response.isSuccessful()) {
                    BaseResponse<List<Delivery>> baseResponse = response.body();
                    if (baseResponse != null && baseResponse.isSuccess()) {
                        List<Delivery> deliveries = baseResponse.getData();

                        if (currentPage == 1) {
                            deliveryAdapter = new DeliveryAdapter(deliveries);
                            recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
                            recyclerView.setAdapter(deliveryAdapter);
                        } else {
                            deliveryAdapter.addDeliveries(deliveries);
                        }

                        if (deliveries != null && deliveries.size() < PAGE_SIZE) isLastPage = true;
                        else currentPage++;
                    } else {
                        isLastPage = true;
                    }
                } else {
                    Log.e("API Error", "Response not successful: " + response.code());
                }
            }

            @Override
            public void onFailure(@NonNull Call<BaseResponse<List<Delivery>>> call, @NonNull Throwable t) {
                isLoading = false;
                Log.e("API Error", "Failed to fetch deliveries: " + t.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/simple/tracking/admin/fragment/DeliveryFragment.kt
package com.simple.tracking.admin.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.cardview.widget.CardView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.simple.tracking.admin.adapter.DeliveryAdapter
import com.simple.tracking.R
import com.simple.tracking.admin.activity.delivery.AdminCreateDeliveryDetailActivity
import com.simple.tracking.admin.activity.user.AdminCreateUserActivity
import com.simple.tracking.model.Delivery

class DeliveryFragment : Fragment() {

    private lateinit var deliveryAdapter: DeliveryAdapter
    private lateinit var recyclerView: RecyclerView
    private lateinit var btnAddDelivery: CardView

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_admin_delivery, container, false)

        recyclerView = view.findViewById(R.id.recyclerView)
        btnAddDelivery = view.findViewById(R.id.btn_add_delivery)

        btnAddDelivery.setOnClickListener {
            val userCreate = Intent(requireContext(), AdminCreateDeliveryDetailActivity::class.java)
            userCreate.putExtra("MENU_NAME", "User")
            startActivity(userCreate)
        }

        val deliveryList = listOf(
            Delivery("Company Name 1", "Address Company Name 1"),
            Delivery("Company Name 2", "Address Company Name 2"),
            Delivery("Company Name 3", "Address Company Name 3"),
            Delivery("Company Name 4", "Address Company Name 4"),
            Delivery("Company Name 5", "Address Company Name 5"),
            Delivery("Company Name 6", "Address Company Name 6"),
            Delivery("Company Name 7", "Address Company Name 7"),
            Delivery("Company Name 8", "Address Company Name 8"),
            Delivery("Company Name 9", "Address Company Name 9"),
            Delivery("Company Name 10", "Address Company Name 10"),
            Delivery("Company Name 11", "Address Company Name 11"),
            Delivery("Company Name 12", "Address Company Name 12"),
            Delivery("Company Name 13", "Address Company Name 13"),
            Delivery("Company Name 14", "Address Company Name 14")
        )

        deliveryAdapter = DeliveryAdapter(deliveryList)
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = deliveryAdapter

        return view
    }
}